sys-boot/{udk,refind}: Simplify installation and bump to new version

sveyret opened the pull request:
There are actually 2 linked modifications in this PR. The first one simplifies installation of UDK (see referenced bug for more information) and the other one upgrade the refind package.

sveyret commented:
Work In Progress: I realized that I forgot to care of no symlink libs, as treated in https://bugs.gentoo.org/627744. I will upload a correction in a few days.

sveyret commented:
Work done. Thank you.

mgorny commented on sys-boot/udk/udk-2018-r1.ebuild:
```diff
@@ -0,0 +1,204 @@
+
+src_install() {
+	exeinto "/usr/bin"
+	doexe udk-workspace
```
Why are you reinventing `dobin`?

sveyret replied:
Because I forgot about it ! ;)

mgorny commented on sys-boot/udk/udk-2018-r1.ebuild:
```diff
@@ -0,0 +1,204 @@
+		exeinto "/usr/share/${P}/${filename%/*}"
+		doexe "${filename}"
+		rm -f "${filename}"
+	done < <(find . -executable -type f -print0)
```
`doins -r` should preserve executable bits, you don't need to hack it around.

sveyret replied:
Actually, I did the hack because ’doins -r’ did not preserve executable bits, when I did my tests. Is there a flag or something I should give?

mgorny replied:
You are correct, it doesn't. Which is kinda silly.

I think the simplest solution would be to replace this whole block with plain `cp -R -p`.

mgorny commented on sys-boot/refind/refind-0.11.4.ebuild:
```diff
@@ -0,0 +1,197 @@
+
+EAPI=7
+
+inherit toolchain-funcs flag-o-matic multiprocessing
```
Sort.

mgorny commented on sys-boot/refind/refind-0.11.4.ebuild:
```diff
@@ -0,0 +1,197 @@
+
+DESCRIPTION="The UEFI Boot Manager by Rod Smith"
+HOMEPAGE="https://www.rodsbooks.com/refind/"
+
```
Remove this empty line.

mgorny commented on profiles/base/package.use.force:
```diff
@@ -38,6 +38,7 @@ media-tv/kodi system-ffmpeg
 # Sam Jorna <wraeth@gentoo.org> (15 Dec 2016)
 # Known to build unbootable EFI binaries on some systems when built with
 # UDK. Forcing use of gnu-efi until it's more stable.
+# This will not be useful anymore as gnuefi use has been deleted.
```
Do you mean the flag? Not sure whom the comment is intended to, or how it helps.

sveyret replied:
Yes, the use flag has been deleted in the latest version of rEFInd. I added the comment for we know we  can remove this package.use.force when version 0.11.4 of rEFInd is stabilized. Shall I update the comment to put all this information?

mgorny replied:
No need to change this at all. When all versions having this flag are removed, pkgcheck will tell us it's obsolete.

mgorny commented on sys-boot/refind/refind-0.11.4.ebuild:
```diff
@@ -0,0 +1,197 @@
+
+DEPEND="|| ( >=sys-boot/gnu-efi-3.0.2 >=sys-boot/udk-2018-r1 )"
+
+DOCS="NEWS.txt README.txt docs/refind docs/Styles"
```
Use bash array, please.

mgorny commented on sys-boot/refind/refind-0.11.4.ebuild:
```diff
@@ -0,0 +1,197 @@
+DOCS="NEWS.txt README.txt docs/refind docs/Styles"
+
+UDK_ROOT="${EPREFIX}"/usr/share/udk
+[[ -d /usr/share/udk ]] && UDK_ROOT=/usr/share/udk
```
Whatever you're trying to do here, it's definitely wrong. You aren't allowed to access filesystem in global scope.

sveyret replied:
What is the suggested way to manage this? I need to find `udk` installation in order to compile rEFInd. But `udk` may have been installed with a prefix. The way I wanted to manage was to try first searching it using same prefix as rEFInd. If not found, I search it in its default place.
Do you think I should set this variable in `src_prepare` or do you think I should completely forget about prefix (and maybe display a warning in `udk` installation)?

mgorny replied:
You should use EPREFIX unconditionally.

mgorny commented on sys-boot/refind/refind-0.11.4.ebuild:
```diff
@@ -0,0 +1,197 @@
+	elif use amd64; then
+		export EFIARCH=x64
+		export BUILDARCH=x86_64
+	else
```
Again, this shouldn't ever happen.

mgorny commented on sys-boot/refind/refind-0.11.4.ebuild:
```diff
@@ -0,0 +1,197 @@
+	default
+
+	# bug 598647 - PIE not supported
+	sed -e 's:CFLAGS          =:& -fno-PIE:' -i Make.common || die
```
Don't match exact whitespace, it's fragile. Can't you just e.g. `/^CFLAGS/s:$: -fno_PIE:`?

sveyret replied:
I'd never think of this notation. Now I want to review all my `sed`s! ;)

mgorny commented on sys-boot/refind/refind-0.11.4.ebuild:
```diff
@@ -0,0 +1,197 @@
+	sed -e 's:CFLAGS          =:& -fno-PIE:' -i Make.common || die
+
+	# Update Makefile
+	sed -e "s:\(EDK2BASE=\).*$:\1${UDK_ROOT}:" \
```
Since it's a Makefile, substitute a make variable for the path, and pass the variable to make. This will prevent it from exploding when path accidentally matches your sed chars.

Also, this is really asking for a patch file instead. Possibly replace both seds with a patch.

sveyret replied:
Depending on what you answer above, UDK_ROOT may not be a static value, and so could not be inside a patch. Or I would need to `sed` the patched file…

sveyret replied:
So `UDK_ROOT` depends on `EPREFIX`, then patch is not a good solution here.

sveyret replied:
OK for make variable.

mgorny replied:
In fact, can't you just pass EDK2BASE to emake?

sveyret replied:
I have not tested, but I'm not sure it would work. Actually, EDK2BASE is exported in the Makefile, probably because it is needed in some calls, and simply passing it as a variable to Makefile could not fit the requirements.

sveyret replied:
After testing, this work.

mgorny commented on sys-boot/refind/refind-0.11.4.ebuild:
```diff
@@ -0,0 +1,197 @@
+		-i Makefile || die "Could not update Makefile"
+
+	# Update fs targets depending on uses
+	local fs
```
You can declare multiple vars in a single `local`

mgorny commented on sys-boot/refind/refind-0.11.4.ebuild:
```diff
@@ -0,0 +1,197 @@
+	for fs in ${FS_USE}; do
+		fs=${fs#+}
+		if use "${fs}"; then
+			fs_names="${fs_names}${fs} "
```
Use a bash array instead of that whitespace game.

mgorny commented on sys-boot/refind/refind-0.11.4.ebuild:
```diff
@@ -0,0 +1,197 @@
+			fs_names="${fs_names}${fs} "
+			gnuefi_fs="${gnuefi_fs}${fs}_gnuefi "
+		else
+			sed -e "s:\(\s*RefindPkg/filesystems/${fs}\.inf\):#\1:" \
```
Why are you commenting it out instead of deleting it?

mgorny commented on sys-boot/refind/refind-0.11.4.ebuild:
```diff
@@ -0,0 +1,197 @@
+LICENSE="BSD GPL-2 GPL-3 FDL-1.3"
+SLOT="0"
+KEYWORDS="~amd64 ~x86"
+FS_USE="btrfs +ext2 +ext4 hfs +iso9660 ntfs reiserfs"
```
What is exactly the purpose of that whole fs hackery? Does it have any real gain, or does it merely make the result more 'pretty' at the cost of adding lot of useless complexity to the ebuild?

sveyret replied:
Each file system support can be built or not, so the need to have there name in the use flags. As there is a loop in `src_prepare` to decide which file system should be built, it is easier to separate the fs use flags from other ones. I can then use `FS_USE` in `src_prepare`.

mgorny replied:
But is there a value in not building all of them unconditionally?

sveyret replied:
Well, here are the values I can think about:
- time and space saving (even if it is not that big);
- it is expected by upstream;
- there used to be failing fs building in previous versions, so users may want to disable compilation of these fs if not needed for them;
- gentoo users like to control everything they build on their system… ;)

mgorny replied:
If upstream expects it, then why do you need to patch it in?

sveyret replied:
If we want to build only selected file systems using upstream `Makefile`, we have to directly call the `Makefile` which is in the `filesystem` directory, either providing the variable containing the wanted fs, or calling `make` for each fs one after the other. But this requires to select the way the compilation is made (either `gnuefi` or `udk`) by ourselves, because this selection is made only in the main `Makefile`.

For a matter of consistency, I prefer to let the main `Makefile` always choose the way everything is compiled, which then requires to update the fs variable in the `Makefile` of `filesystem`.

One thing I could do anyway is not removing the fs from `Refind.dsc` (which always build all). In this case, for `udk` style building, it would build every fs, but install only selected ones. But the cost of removing the fs is low, and it increase _safetyness_ in case the compilation does not work for a particular unneeded fs.

mgorny replied:
How do the two 'ways' of compiling differ? Is the end result different?

Compilation not working is not an argument. USE flags are not workarounds for sloppy packages. You are supposed to test them all, and make sure all of them work.

sveyret replied:
Yes, the end result is certainly different, because the compilation is not based on the same libraries.

Regarding compilation not working, OK, I understand. So I should not patch `Refind.dsc`.

mgorny replied:
If the result is different, then `||` is not acceptable and you must either settle on one, or provide explicit choice between the two (i.e. USE flags).

As for filesystems, I believe you should remove the flags and just build them unconditionally unless you can either make it work without sed, or find value that really justifies that amount of extra code.

sveyret replied:
OK, I will keep using the `gnuefi` use flag, then. And use Makefile variables only for filesystems.

sveyret replied:
Done.

sveyret commented on sys-boot/udk/udk-2018-r1.ebuild:
```diff
@@ -0,0 +1,204 @@
+		EFIARCH=IA32
+	elif use amd64; then
+		EFIARCH=X64
+	else
```
I was told once that if someone wants to try and force build on another arch, I should not prevent her/him from doing that. That's why this else statement.
Correcting…

sveyret commented on sys-boot/udk/udk-2018-r1.ebuild:
```diff
@@ -0,0 +1,204 @@
+"
+
+pkg_setup() {
+	if use x86; then
```
Except if, as I previously did, someone change the value of ARCH in ebuild… ;)
Anyway, I feel it is “nicer” to use the “use <arch>” than “== ${ARCH}”

sveyret commented on sys-boot/udk/udk-2018-r1.ebuild:
```diff
@@ -0,0 +1,204 @@
+	find ./BaseTools/Source/C -mindepth 1 -maxdepth 1 \! -name 'bin' -exec rm -rf {} \;
+
+	# Create workspace script file
+	cat >>"${S}/udk-workspace" <<-EOF \
```
OK

sveyret commented:
Thank you @mgorny for the review. I am waiting for your answers to my questions (on un-resolved conversation) before sending a new version of the PR.

sveyret commented:
Thank you again, @mgorny. I updated the PR. I kept two conversation non-resolved in case you still want to answer because I did not change anything at these points.

sveyret commented:
@mgorny, I updated the files. I would like to warn you about 3 points:
- I re-wrote a comment on package.use.force, as the `gnuefi` use flag is finally kept.
- The Makefile for refind is updated using a sed, because there finally is only one line to modify.
- I discovered that HFS is not compiling in UDK mode. That's why the `Refind.dsc` is patched to remove this filesystem compilation. An issue is opened upstream (reference is written in the ebuild as comment).

mgorny commented on sys-boot/refind/refind-0.11.4.ebuild:
```diff
@@ -0,0 +1,187 @@
+
+src_configure() {
+	if ! use gnuefi; then
+		. udk-workspace "${UDK_WORKSPACE}" || die
```
Put it in big subshell not to leak anything from `udk-workspace` into ebuild.

mgorny commented on sys-boot/refind/refind-0.11.4.ebuild:
```diff
@@ -0,0 +1,187 @@
+src_compile() {
+	# Ensure environment is set
+	if ! use gnuefi; then
+		. udk-workspace "${UDK_WORKSPACE}" || die
```
Same as above, don't source external files in ebuild environment. Move it to a subshell around `emake` call.

mgorny commented on sys-boot/refind/refind-0.11.4.ebuild:
```diff
@@ -0,0 +1,187 @@
+	for fs in ${FS_USE}; do
+		fs=${fs#+}
+		if use "${fs}"; then
+			fs_names+=("${fs}$(usex gnuefi _gnuefi '')")
```
Remove this usex here, and add below:
```
use gnuefi && fs_names=( "${fs_names[@]/%/_gnuefi}" )

mgorny commented on sys-boot/refind/refind-0.11.4.ebuild:
```diff
@@ -0,0 +1,187 @@
+	doexe refind-install
+	dosym "../share/${P}/refind-install" "/usr/sbin/refind-install"
+
+	dodoc "${S}"/{COPYING.txt,LICENSE.txt,CREDITS.txt}
```
1. S is the current directory, isn't it? Here and below.
2. Why are you installing license files? That's against Gentoo policy.

mgorny commented on sys-boot/refind/refind-0.11.4.ebuild:
```diff
@@ -0,0 +1,187 @@
+	dodoc "${S}"/{COPYING.txt,LICENSE.txt,CREDITS.txt}
+	if use doc; then
+		doman "${S}/docs/man/"*
+		dodoc -r ${DOCS}
```
That's not how you use an array. Also, why are you randomly mixing and splitting direct `dodoc` and `DOCS`? That makes no sense. If you need things split, just inline both, so people aren't confused that `DOCS` doesn't work like DOCS is supposed to work. Also, README should be installed unconditionally.

sveyret commented:
@mgorny: modifications done. Finally, for udk-workspace, I put the call directly in Makefile, and so added a patch instead of the `sed`.

mgorny commented on sys-boot/udk/files/udk-workspace.template:
```diff
@@ -0,0 +1,62 @@
+#!/bin/bash
```
Honestly, I don't really understand why you are using bash and at the same time limiting yourself to ugly POSIX-ish syntax (i.e. minimal `[ ... ]`).

sveyret replied:
Oh yes, I hate this syntax too. But even if this script could be called directly, it must be sourced to work. So I cannot know which shell will really execute it.

mgorny replied:
Indeed. But then, you are using `BASH_SOURCE` in a few places. Though I guess it only really matters when calling the script directly, correct?

sveyret replied:
The way I understood it is that BASH_SOURCE[0] contains the name of the script, be it called directly or sourced, on the contrary to $0… The values are compared to know if script is sourced.

mgorny replied:
I just meant that it works only in bash ;-).

sveyret replied:
Oops! Correct… What would you suggest?

sveyret replied:
Regarding comparison, if this is not bash, the script will not detect that it is not sourced. But well, this is not a big problem. The other 2 places are used to display the name of the script to end user, for a help on script usage.

sveyret replied:
Maybe test if -z BASH_SOURCE[0], then set it to "udk-workspace"?

mgorny replied:
I think it's kinda fine… as long as the `[0]` part doesn't trigger syntax error. Maybe replace it with plain `${BASH_SOURCE}` that should be equivalent.

mgorny commented on sys-boot/udk/udk-2018-r1.ebuild:
```diff
@@ -0,0 +1,136 @@
+	dobin "${WORKDIR}"/script/udk-workspace
+
+	# Use mkdir && cp here as doins does not preserve execution bits
+	mkdir --parent "${D}/usr/share/${P}" || die
```
`-p` is actually more portable; long options are rarely supported outside GNU. `${ED}` for prefix compat.

mgorny commented on sys-boot/refind/refind-0.11.4.ebuild:
```diff
@@ -0,0 +1,179 @@
+
+src_configure() {
+	if ! use gnuefi; then
+		(. udk-workspace "${UDK_WORKSPACE}" || die)
```
This is actually meaningless. You declare variables and then discard them immediately.

sveyret replied:
This is because udk-workspace must be sourced, but here, I only use the file creation side effect.

mgorny replied:
Please add a comment whenever you do something potentially confusing.

sveyret replied:
OK, comment added…

mgorny commented on sys-boot/udk/udk-2018-r1.ebuild:
```diff
@@ -0,0 +1,136 @@
+DOCDIR="${WORKDIR}/Documents"
+
+# EFI pre-build libs
+QA_PREBUILT="
```
Sorry for missing this earlier. Executables are not allowed in `/usr/share`, that tree is for arch-independent stuff. `/usr/lib` is for binaries.

sveyret replied:
Is $(get_libdir) allowed in global context?

mgorny commented on sys-boot/refind/refind-0.11.4.ebuild:
```diff
@@ -0,0 +1,179 @@
+}
+
+src_install() {
+	exeinto "/usr/share/${P}"
```
Same as in the other package, `/usr/share` is wrong here.

sveyret commented:
@xdch47 @mgorny modifications done (or commented). Thank you.

mgorny commented on sys-boot/udk/files/udk-workspace.template:
```diff
@@ -0,0 +1,62 @@
+	local workspace_dir="${PWD}"
+	if [ ! -z "${1}" ]; then
+		workspace_dir="${1}"
+		if [ ! -d "${1}" ]; then
```
Technically speaking, this is unnecessary, since `mkdir -p` returns successfully if the directory exists already.

mgorny commented on sys-boot/udk/files/udk-workspace.template:
```diff
@@ -0,0 +1,62 @@
+		fi
+	fi
+
+	pushd "${workspace_dir}" >>/dev/null
```
I think it's customary to use `>/dev/null`. There's no technical value in appending to it ;-).

mgorny commented on sys-boot/udk/files/udk-workspace.template:
```diff
@@ -0,0 +1,62 @@
+	fi
+
+	pushd "${workspace_dir}" >>/dev/null
+	if [ ! -d Conf ] && ! mkdir Conf; then
```
`mkdir -p` would probably be more readable here (without the check).

mgorny commented on sys-boot/udk/files/udk-workspace.template:
```diff
@@ -0,0 +1,62 @@
+	popd >>/dev/null
+}
+
+if [ "${1}" = "-h" ] || [ "${1}" = "--help" ]; then
```
Unless I'm mistaken, this can match `${1}` of sourcing script.

sveyret replied:
Correct… :disappointed: 
Should I only accept these parameters on direct calls?

sveyret replied:
I think I'll keep this and hope users will clear parameters before calling the script. For almost every usage, this should work correctly. It may fail if the user source the script inside a script taking parameters, and in this case, he will have to take care about those parameters.

mgorny commented on sys-boot/udk/files/udk-workspace.template:
```diff
@@ -0,0 +1,62 @@
+if [ "${1}" = "-h" ] || [ "${1}" = "--help" ]; then
+	show_usage
+	# Script may have been sourced
+	return 0 2>>/dev/null
```
Hmm, I didn't know you can `return` from sourcing. Is this portable? If you're not sure, then it would probably be better to wrap it into `main()` function, and `return` from that.

sveyret replied:
I would not be able to give an error code on script exit…

mgorny replied:
Why not? Just `return` from `main()` with an exit code, and the script's exit code will be the exit code of last command executed, i.e. `main`.

sveyret replied:
Good idea, but then, I will not be able to unset main() and other functions. Should I prefix them and not unset them?

mgorny replied:
I've checked with POSIX and `return` is valid here.

sveyret replied:
Great, so I'll keep as is.

sveyret commented:
@mgorny Changes done.

mgorny commented on sys-boot/udk/udk-2018-r1.ebuild:
```diff
@@ -0,0 +1,136 @@
+	mkdir "${WORKDIR}/script"
+	sed -e "s:{EDK_BASE}:${EPREFIX}/usr/$(get_libdir)/${P}:" \
+		"${FILESDIR}"/udk-workspace.template \
+		>"${WORKDIR}"/script/udk-workspace || die "Failed to build udk-workspace"
```
We normally use `${T}` for this. There's really no need to create extra directories.

mgorny commented on sys-boot/udk/udk-2018-r1.ebuild:
```diff
@@ -0,0 +1,136 @@
+	dobin "${WORKDIR}"/script/udk-workspace
+
+	# Use mkdir && cp here as doins does not preserve execution bits
+	mkdir -p "${ED}/usr/$(get_libdir)/${P}" || die
```
Why get_libdir? You aren't going to build his multilib, are you? Then it's just needless complexity.

sveyret replied:
So I shall put everything in lib? Isn't this reserved for 32 bits? Especially now with profile 17.1? Or should I use `/usr/libexec`?

mgorny replied:
Yes. It isn't.

sveyret commented:
@mgorny Done.

mgorny commented on sys-boot/refind/refind-0.11.4.ebuild:
```diff
@@ -0,0 +1,180 @@
+src_install() {
+	exeinto "/usr/lib/${P}"
+	doexe refind-install
+	dosym "../share/${P}/refind-install" "/usr/sbin/refind-install"
```
Wrong path.

mgorny commented on sys-boot/refind/refind-0.11.4.ebuild:
```diff
@@ -0,0 +1,180 @@
+		elog "A sample configuration can be found at"
+		elog "${EROOT}/usr/lib/${P}/refind/refind.conf-sample"
+	else
+		if ! version_is_at_least "0.10.3" "${REPLACING_VERSIONS}"; then
```
'Command not found'

sveyret commented:
@mgorny 

> I have to say I'm dissatisfied that you missed them while testing.

I understand. Sorry for that. I don't know how I could miss it, especially the versionator related issue, as I removed this deprecated eclass a very long time ago and did a lot of tests since… :disappointed_relieved: 

> I'm not going to block this merge on this but please address it in a followup:

What's good is that you make me learn a lot. :smile: I did not know there was a way to see the *FLAGS respect. It made me realize that there were some problems with LDFLAGS. These are now corrected.

By the way, there also are a lot of “jobserver unavailable: using -j1” QA notices when compiling `refind` with `udk` (`gnuefi` flag off), but I did not find the reason why. I did not find any remaining `-j1` in `refind` compilation for `udk`. I will open an issue upstream.

Regarding the bad link hack of `udk`, I removed it. If a user wants it for a specific build, he will have to create it with the good target.

xdch47 commented:
it is important that the filesystems targets are **NOT** compiled in parallel.
the makefile is a mess and basically working like that:
- setting the `FSTYTPE` define
- delete `fsw_efi.o` (if exists)
- compile `fsw_efi.c` and link the library,
- repeat those steps for an other filesystem.

I suggest to patch the `filesystems/Makefile` as well, otherwise the `udk` compilation will fail as well, as soon as `+$(MAKE)` (fixes the jobserver issue) on the toplevel makefiles are added.  -- the other solution would be to add a `-j1` to the other `emake`.
```
diff --git a/filesystems/Makefile b/filesystems/Makefile
index 99452ed..9bc984d 100644
--- a/filesystems/Makefile
+++ b/filesystems/Makefile
@@ -5,6 +5,7 @@
 # name of the driver to be built. This is done because of a dependency
 # in the fsw_efi.c file on the filesystem type; this file must be
 # recompiled for each new filesystem built.
+.NOTPARALLEL:
 
 # This program is licensed under the terms of the GNU GPL, version 3,
 # or (at your option) any later version.
```

sveyret commented:
Thanks @xdch47, I'll work on that asap. When I think of it, I wonder if this parallel problem would not be the reason why `hfs` is not working with `udk`.

sveyret commented:
Thanks @xdch47. When I think of it, I wonder if this parallel problem would not be the reason why `hfs` is not working with `udk`.
But unfortunately, I don't think there is anything to do with Refind makefiles, as `udk` build is made using the `build` command, and not using a `Makefile`. `Tianocore` target is using the `Makefile`, but it is an old version of `udk` which is not the one I use.

xdch47 commented:
okay,
did not tried udk build before - just had a look on the makefile.

just tested the udk build, I had to remove `-Werror` (gcc 9.1) [as mention in the BUILDING.txt for OS X]
```
sed 
	-e "s:-Werror::" \
	-i "${UDK_WORKSPACE}/Conf/tools_def.txt"
```

furthermore add `+` remove the -j1 QA notice (see complete patch):
```diff --git a/Makefile b/Makefile
index ea57f16..930096c 100644
--- a/Makefile
+++ b/Makefile
@@ -161,8 +161,9 @@ fs_edk2: build_edk2
 	done
 
 build_edk2: $(EDK2BASE)/RefindPkg
-	cd $(EDK2BASE) && \
-	. ./edksetup.sh BaseTools && \
+	+cd $(EDK2BASE) && \
+	. udk-workspace && \
+	build -a $(UC_ARCH) -p MdePkg/MdePkg.dsc && \
 	build -a $(UC_ARCH) -p RefindPkg/RefindPkg.dsc
 	mkdir -p ./drivers_$(FILENAME_CODE)
 
diff --git a/filesystems/Makefile b/filesystems/Makefile
index 99452ed..9bc984d 100644
--- a/filesystems/Makefile
+++ b/filesystems/Makefile
@@ -5,6 +5,7 @@
 # name of the driver to be built. This is done because of a dependency
 # in the fsw_efi.c file on the filesystem type; this file must be
 # recompiled for each new filesystem built.
+.NOTPARALLEL:
 
 # This program is licensed under the terms of the GNU GPL, version 3,
 # or (at your option) any later version.
```

xdch47 commented on sys-boot/refind/refind-0.11.4.ebuild:
```diff
@@ -0,0 +1,177 @@
+}
+
+src_install() {
+	exeinto "/usr/lib/${P}"
```
is there actually a reason why to use packagename-version `${P}` instead of just the packagename `${PN}` ?

sveyret replied:
Well, the reason is that, in the future, we may want to slot the package, because versions are not always fully ascending compatible. Note that I create a symlink from `${PN}` to `${P}` anyway.

sveyret replied:
Sorry, I mixed up the `udk` package and the `refind` one. No, there is no good reason for this one. I will update this.

sveyret commented:
@xdch47 Thanks a lot for your suggestions! Indeed, with the `+`, I don't have all those jobserver issues anymore, and thanks to the `-Werror` tip, it is now also possible to build the `hfs` filesystem, which I had disabled due to errors.
@mgorny I think the latest version should be rather good… :smile:

xdch47 commented on sys-boot/refind/refind-0.11.4.ebuild:
```diff
@@ -0,0 +1,168 @@
+EAPI=7
+
+inherit flag-o-matic multiprocessing toolchain-funcs
+
```
why `flag-o-matic` and `multiprocessing`?

sveyret replied:
`multiprocessing` is used for `makeopts_jobs`, `flag-o-matic` was used in previous version and is now useless. I don't know how or even if it is possible, but it would be good to have `repoman` detect those cases…

xdch47 commented on sys-boot/udk/udk-2018-r1.ebuild:
```diff
@@ -0,0 +1,147 @@
+PYTHON_COMPAT=( python2_7 )
+PYTHON_REQ_USE="sqlite"
+
+inherit flag-o-matic multiprocessing python-single-r1 toolchain-funcs
```
why `flag-o-matic` ?

sveyret replied:
Done

mgorny commented on sys-boot/udk/udk-2018-r1.ebuild:
```diff
@@ -0,0 +1,147 @@
+	rm "${S}/EmulatorPkg/Unix/Host/X11IncludeHack"
+
+	# Create workspace script file
+	mkdir "${T}/script"
```
Why do you insist on creating a subdirectory? oO

sveyret replied:
Because I feel more comfortable this way. If we need to put something else in this temporary directory, it will not be mixed with the script. But if you don't like it, no problem, I'll remove it.

mgorny replied:
It's a temporary directory for you to use, so please avoid introducing unnecessary subdirectories.

mgorny replied:
Or if you're up for something really crazy, you can do something like:

    newins - udk-workspace < <(sed ...)

sveyret replied:
I will refrain… :wink:

mgorny commented on sys-boot/udk/udk-2018-r1.ebuild:
```diff
@@ -0,0 +1,147 @@
+	find ./BaseTools/Source/C -mindepth 1 -maxdepth 1 \! -name 'bin' -exec rm -rf {} \;
+
+	# Upsteam hack (symbolic link) should only be created if needed
+	rm "${S}/EmulatorPkg/Unix/Host/X11IncludeHack"
```
||die

mgorny commented on sys-boot/udk/udk-2018-r1.ebuild:
```diff
@@ -0,0 +1,147 @@
+		-i "BaseTools/Conf/target.template" || die "Failed to configure target file"
+
+	# Clean unneeded files
+	find . -name '*.bat' -o -name '*.exe' -exec rm -f {} \;
```
`find` needs ||die too
